### ISSUE TITLE
[Optymalizacja] ReplayBuffer.shift() i TensorSync

### ISSUE THREAD

**Issue by ArgonZapan** [ArgonZapan/Checkers_vibe#11]
## Optymalizacja — Performance issues

**Autor:** Jarvis (AI QA)

---

### Lokalizacja: `server/ai/buffer.js`, linia ~18

```js
add(sample) {
  this.buffer.push(sample);
  while (this.buffer.length > this.maxSize) {
    this.buffer.shift();  // ❌ O(n) — kopiuje całą tablicę
  }
}
```

### Problem:
`Array.shift()` usuwa pierwszy element i przesuwa WSZYSTKIE pozostałe elementy w pamięci. Przy 10000 elementów i częstym add(), to może być wolne.

### Rozwiązanie — indeks cykliczny:
```js
this.buffer = new Array(maxSize);
this.head = 0;

add(sample) {
  this.buffer[this.head] = sample;
  this.head = (this.head + 1) % this.maxSize;
  this.buffer.length = Math.min(this.head + 1, this.maxSize);
}
```

---

### Lokalizacja: `server/ai/model.js`, linia ~35

```js
await model.fit(xTensor, [...], {
  epochs,
  batchSize: Math.min(batch.length, 64),  // ❌ maks 64
  verbose: 0
});
```

### Problem:
Batch size 64 z 5 epochami na 256 samplach = ~20 gradient updates. Większy batch = szybsze training bo mniej overheadu per sample.

### Rozwiązanie:
```js
batchSize: Math.min(batch.length, 256),  // lub cały batch
```

Lub lepiej — trenuj na całym batchu:
```js
batchSize: batch.length,
epochs: Math.max(1, Math.ceil(20 / Math.sqrt(batch.length))),
```

---

### Lokalizacja: `server/ai/model.js`, linia ~12

```js
boards.push(tensor.arraySync()[0]);  // ❌ sync call
```

### Problem:
`arraySync()` blokuje główny wątek JavaScript. Dla 64-sample batch to ~64 × 257 × 4 bytes = ~65KB danych, ale synchronicznie.

### Rozwiązanie:
```js
// Zamiast arraySync():
const data = await tensor.data();  // async, nie blokuje
boards.push(Array.from(data));
```

---

### Lokalizacja: `server/ai/trainer.js`, linia ~217

```js
const lossWhite = await train(this.modelWhite, batch.filter(s => s.turn === 1), 5);
const lossBlack = await train(this.modelBlack, batch.filter(s => s.turn === -1), 5);
```

### Problem:
Tworzysz 2 nowe tablice przez filtrowanie przy każdym treningu. Przy 1000 gier = 1000 filtrowań.

### Rozwiązanie:
Rozdziel buffer na dwa — `bufferWhite` i `bufferBlack` — wtedy nie trzeba filtrować przy treningu.

---

## 📋 Podsumowanie

| Optymalizacja | Impact | Priorytet |
|--------------|--------|----------|
| ReplayBuffer.shift() → cykliczny | 🟠 Średni | Wysoki |
| batchSize 64 → 256 | 🟢 Mały gain | Średni |
| arraySync() → async data() | 🟠 Średni | Średni |
| Batch filter split | 🟢 Kosmetyka | Niski |

---

*Optymalizacja przez Jarvis AI (QA).*
